enterprise/coderd/prebuilds/claim_test.go

```diff
@@ -0,0 +1,367 @@
+				require.NoError(t, reconciler.ReconcilePreset(ctx, *ps))
+			}
+
+			// Given: a set of running, eligible prebuilds eventually starts up.
+			runningPrebuilds := make(map[uuid.UUID]database.GetRunningPrebuiltWorkspacesRow, desiredInstances*presetCount)
+			require.Eventually(t, func() bool {
+				rows, err := spy.GetRunningPrebuiltWorkspaces(ctx)
+				require.NoError(t, err)
+
+				for _, row := range rows {
+					runningPrebuilds[row.CurrentPresetID.UUID] = row
+
+					if !tc.markPrebuildsClaimable {
+						continue
+					}
+
+					agents, err := db.GetWorkspaceAgentsInLatestBuildByWorkspaceID(ctx, row.ID)
+					require.NoError(t, err)
+
+					// Workspaces are eligible once its agent is marked "ready".
+					for _, agent := range agents {
+						require.NoError(t, db.UpdateWorkspaceAgentLifecycleStateByID(ctx, database.UpdateWorkspaceAgentLifecycleStateByIDParams{
+							ID:             agent.ID,
+							LifecycleState: database.WorkspaceAgentLifecycleStateReady,
+							StartedAt:      sql.NullTime{Time: time.Now().Add(time.Hour), Valid: true},
+							ReadyAt:        sql.NullTime{Time: time.Now().Add(-1 * time.Hour), Valid: true},
+						}))
+					}
+				}
+
+				t.Logf("found %d running prebuilds so far, want %d", len(runningPrebuilds), expectedPrebuildsCount)
+
+				return len(runningPrebuilds) == expectedPrebuildsCount
+			}, testutil.WaitSuperLong, testutil.IntervalSlow)
+
+			// When: a user creates a new workspace with a preset for which prebuilds are configured.
+			workspaceName := strings.ReplaceAll(testutil.GetRandomName(t), "_", "-")
+			params := database.ClaimPrebuiltWorkspaceParams{
+				NewUserID: user.ID,
+				NewName:   workspaceName,
+				PresetID:  presets[0].ID,
+			}
+			userWorkspace, err := userClient.CreateUserWorkspace(ctx, user.Username, codersdk.CreateWorkspaceRequest{
+				TemplateVersionID:       version.ID,
+				Name:                    workspaceName,
+				TemplateVersionPresetID: presets[0].ID,
+			})
+
+			require.NoError(t, err)
+			coderdtest.AwaitWorkspaceBuildJobCompleted(t, userClient, userWorkspace.LatestBuild.ID)
+
+			// Then: a prebuild should have been claimed.
+			require.EqualValues(t, spy.claims.Load(), 1)
+			require.NotNil(t, spy.claims.Load())
+			require.EqualValues(t, *spy.claimParams.Load(), params)
+
+			if !tc.expectPrebuildClaimed {
+				require.Nil(t, spy.claimedWorkspace.Load())
+				return
+			}
+
+			require.NotNil(t, spy.claimedWorkspace.Load())
+			claimed := *spy.claimedWorkspace.Load()
+			require.NotEqual(t, claimed, uuid.Nil)
+
+			// Then: the claimed prebuild must now be owned by the requester.
+			workspace, err := spy.GetWorkspaceByID(ctx, claimed.ID)
+			require.NoError(t, err)
+			require.Equal(t, user.ID, workspace.OwnerID)
+
+			// Then: the number of running prebuilds has changed since one was claimed.
+			currentPrebuilds, err := spy.GetRunningPrebuiltWorkspaces(ctx)
+			require.NoError(t, err)
+			require.NotEqual(t, len(currentPrebuilds), len(runningPrebuilds))
```

Comment:
Self-review:

I guess it's better to use:
`require.Equal(t, len(currentPrebuilds), len(runningPrebuilds))-1`